### PR TITLE
Add dropwizard metrics

### PR DESCRIPTION
**Pull request by slfritchie**
Replacement for PR #424.  Contains corrections requested in 424's review process.

### CODE REVIEW

File: infrastructure/src/main/java/org/corfudb/infrastructure/AbstractServer.java [hunk 1]
```diff
@@ -36,10 +39,66 @@ public AbstractServer() {
      * @param r   The router that took in the message.
      */
     public void handleMessage(CorfuMsg msg, ChannelHandlerContext ctx, IServerRouter r) {
+        boolean isEnabled = MetricsUtils.isMetricsCollectionEnabled();
+        Timer.Context context = null;
+
         if (isShutdown()) return;
+        switch (msg.getMsgType()) {
```
Comment (dahliamalkhi) [inline infrastructure/src/main/java/org/corfudb/infrastructure/AbstractServer.java]: I am not happy with this long/manual switch. Once again, service components should be able to register themselves to the metering infrastructure. This is more modular, and scopes everything relevant to a component within that file. Advantages include, for example, that there is no need to change in multiple places when components change. 

Similarly to the way different message types are handled by the different servers, AbstractServer should have a generic function like getHandler(). It can return an object similar to CorfuMsgHandler, which has methods  'getTimeContext(msg-type..)' . And so on.

File: infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java [hunk 2]
```diff
@@ -205,6 +205,11 @@ public static void main(String[] args) {
 
         // Create a common Server Context for all servers to access.
         serverContext = new ServerContext(opts, router);
+        String mpBase = "corfu.server.base.";
```
Comment (dahliamalkhi) [inline infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java]: Why is this here, and not like other components?

File: infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java [hunk 3]
```diff
@@ -102,19 +107,26 @@
...
             .build();
 
     /** Handler for this server */
     @Getter
     private CorfuMsgHandler handler = new CorfuMsgHandler()
             .generateHandlers(MethodHandles.lookup(), this);
 
+    @Getter
+    static private Timer timerSeqReq;
+    private Counter counterTokenSum;
```
Comment (dahliamalkhi) [inline infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java]: should this and the next counter be static?

File: infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java [hunk 4]
```diff
@@ -235,9 +254,13 @@ public synchronized void tokenRequest(CorfuPayloadMsg<TokenRequest> msg,
                                           ChannelHandlerContext ctx, IServerRouter r) {
         TokenRequest req = msg.getPayload();
 
+        boolean isEnabled = MetricsUtils.isMetricsCollectionEnabled();
```
Comment (dahliamalkhi) [inline infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java]: once the above is implemented, we should have a per-server metric-handler(msg ...) , which should handle this.

File: runtime/src/main/java/org/corfudb/runtime/object/CorfuCompileProxy.java [hunk 5]
```diff
@@ -120,6 +138,20 @@ public CorfuCompileProxy(CorfuRuntime rt, UUID streamID, Class<T> type, Object[]
     @Override
     public <R> R access(ICorfuSMRAccess<R, T> accessMethod,
                         Object[] conflictObject) {
+        R value;
+        boolean isEnabled = MetricsUtils.isMetricsCollectionEnabled();
```
Comment (dahliamalkhi) [inline runtime/src/main/java/org/corfudb/runtime/object/CorfuCompileProxy.java]: Once again, I don't understand why obtain the isEnabled boolean from MetricUtils, only to send it back to MetricUtils.getConditionalContext(). Same below in all instances.

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **d8eca02**.

**Throughput**: _Average_: 38.4615384615, _Standard Deviation_: 9.43492191167
**Latency**: _Average_: 253.028, _Standard Deviation_: 116.26327544

 An interactive dashboard with Pull Request Performance Metrics is available at:
[Pull Request #444 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/5OHzhAGLXdP1ogIP5FyGd2vz2OwElkwK)

File: infrastructure/src/main/java/org/corfudb/infrastructure/AbstractServer.java [hunk 6]
```diff
@@ -34,7 +37,8 @@ public AbstractServer() {
      */
     public void handleMessage(CorfuMsg msg, ChannelHandlerContext ctx, IServerRouter r) {
         if (isShutdown()) return;
-        if (!getHandler().handle(msg, ctx, r)) {
+        boolean isMetricsEnabled = MetricsUtils.isMetricsCollectionEnabled();
+        if (!getHandler().handle(msg, ctx, r, isMetricsEnabled)) {
```
Comment (dahliamalkhi) [inline infrastructure/src/main/java/org/corfudb/infrastructure/AbstractServer.java]: you are passing a new flag to all handlers (isMetricEnabled), which I don't see used even once.

I see one place -- SequencerServer -- which needs it. As it turns out, SequencerServer is grabbing this directly from MetricsUtils.

The question is: Should we pass this parameter to all handlers, so that we can avoid that single usage, or the other way around? My sense is that perhaps we shouldn't pass it.

File: infrastructure/src/main/java/org/corfudb/infrastructure/CorfuMsgHandler.java [hunk 7]
```diff
@@ -93,23 +97,48 @@ public CorfuMsgHandler generateHandlers(final MethodHandles.Lookup caller, final
...
+                        } else {
+                            t = null;
                         }
+                        // Now create the lambda that wraps the lambda-like-thing that's
+                        // stored in 'h' and insert it into the handlerMap.
+                        handlerMap.put(a.type(),
+                                (CorfuMsg msg, ChannelHandlerContext ctx, IServerRouter r, boolean isMetricsEnabled) -> {
+                                    Timer.Context timerCxt = MetricsUtils.getConditionalContext(
+                                            t != null && isMetricsEnabled, t);
+                                    try {
```
Comment (dahliamalkhi) [inline infrastructure/src/main/java/org/corfudb/infrastructure/CorfuMsgHandler.java]: should we turn this to try-with-resource?

File: infrastructure/src/main/java/org/corfudb/infrastructure/DataStore.java [hunk 8]
```diff
@@ -36,6 +36,7 @@
 
     private final Map<String, Object> opts;
     private final boolean isPersistent;
+    @Getter
```
Comment (dahliamalkhi) [inline infrastructure/src/main/java/org/corfudb/infrastructure/DataStore.java]: is this still needed?

File: infrastructure/src/main/java/org/corfudb/infrastructure/DataStore.java [hunk 9]
```diff
@@ -67,6 +68,7 @@ public DataStore(Map<String, Object> opts) {
     private LoadingCache<String, String> buildMemoryDS() {
         LoadingCache<String, String> cache = Caffeine
                 .newBuilder()
+                .recordStats()
```
Comment (dahliamalkhi) [inline infrastructure/src/main/java/org/corfudb/infrastructure/DataStore.java]: should this be conditioned on metrics collection enabled?

File: infrastructure/src/main/java/org/corfudb/infrastructure/DataStore.java [hunk 10]
```diff
@@ -80,6 +82,7 @@ public DataStore(Map<String, Object> opts) {
      */
     private LoadingCache<String, String> buildPersistentDS() {
         LoadingCache<String, String> cache = Caffeine.newBuilder()
+                .recordStats()
```
Comment (dahliamalkhi) [inline infrastructure/src/main/java/org/corfudb/infrastructure/DataStore.java]: same

File: infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java [hunk 11]
```diff
@@ -139,11 +143,15 @@ public LogUnitServer(ServerContext serverContext) {
                 .maximumWeight(maxCacheSize)
                 .removalListener(this::handleEviction)
                 .writer(batchWriter)
+                .recordStats()
```
Comment (dahliamalkhi) [inline infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java]: same

File: infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java [hunk 12]
```diff
@@ -139,11 +143,15 @@ public LogUnitServer(ServerContext serverContext) {
...
                 .build(this::handleRetrieval);
 
         // Trim map is set to empty on start
         // TODO: persist trim map - this is optional since trim is just a hint.
         trimMap = new ConcurrentHashMap<>();
+
+        MetricRegistry metrics = serverContext.getMetrics();
+        MetricsUtils.addCacheGauges(metrics, metricsPrefix + "cache.", dataCache);
```
Comment (dahliamalkhi) [inline infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java]: same

File: infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java [hunk 13]
```diff
@@ -102,19 +104,25 @@
 
     private final long maxConflictCacheSize = 10_000;
     private long maxConflictWildcard = -1L;
+    @Getter
```
Comment (dahliamalkhi) [inline infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java]: again, is this still used anywhere?

File: infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java [hunk 14]
```diff
@@ -102,19 +104,25 @@
...
     private final Cache<Integer, Long>
             conflictToGlobalTailCache = Caffeine.newBuilder()
             .maximumSize(maxConflictCacheSize)
             .removalListener((Integer K, Long V, RemovalCause cause) -> {
                 maxConflictWildcard = Math.max(V, maxConflictWildcard);
             })
+            .recordStats()
```
Comment (dahliamalkhi) [inline infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java]: conditional?

File: infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java [hunk 15]
```diff
@@ -125,6 +133,11 @@ public SequencerServer(ServerContext serverContext) {
         } else {
             globalLogTail.set(initialToken);
         }
+
+        MetricRegistry metrics = serverContext.getMetrics();
+        counterTokenSum = metrics.counter(metricsPrefix + "token-sum");
+        counterToken0 = metrics.counter(metricsPrefix + "token-query");
+        addCacheGauges(metrics, metricsPrefix + "conflict.cache.", conflictToGlobalTailCache);
```
Comment (dahliamalkhi) [inline infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java]: conditional?

File: infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java [hunk 16]
```diff
@@ -230,14 +243,19 @@ public void handleTokenQuery(CorfuPayloadMsg<TokenRequest> msg,
...
      */
-    @ServerHandler(type=CorfuMsgType.TOKEN_REQ)
+    @ServerHandler(type=CorfuMsgType.TOKEN_REQ, opTimer=metricsPrefix + "token-req")
     public synchronized void tokenRequest(CorfuPayloadMsg<TokenRequest> msg,
-                                          ChannelHandlerContext ctx, IServerRouter r) {
+                                          ChannelHandlerContext ctx, IServerRouter r,
+                                          boolean isMetricsEnabled) {
         TokenRequest req = msg.getPayload();
 
+        boolean isEnabled = MetricsUtils.isMetricsCollectionEnabled();
```
Comment (dahliamalkhi) [inline infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java]: this is the only place I saw where the conditional should have been used, and isn't

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **8964176**.

**Throughput**: _Average_: 35.7142857143 [executions/sec], _Standard Deviation_: 9.2537912804
**Latency**: _Average_: 272.086 [ms], _Standard Deviation_: 108.505302193

 An interactive dashboard with Pull Request Performance Metrics is available at:
[Pull Request #444 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/I7bGrqCf8k7XF7lL0w3DH1khFP1EEd74)

File: infrastructure/src/main/java/org/corfudb/infrastructure/DataStore.java (hunk 9 above)
Comment (slfritchie) [inline infrastructure/src/main/java/org/corfudb/infrastructure/DataStore.java]: I'm not aware of Caffeine providing a way to disable/enable stats other than this builder function.  If we don't enable it here, there's no choice later.

File: infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java (hunk 16 above)
Comment (slfritchie) [inline infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java]: Sorry, that was a leftover that should have been removed, thanks.

File: runtime/src/main/java/org/corfudb/runtime/object/CorfuCompileProxy.java (hunk 5 above)
Comment (slfritchie) [inline runtime/src/main/java/org/corfudb/runtime/object/CorfuCompileProxy.java]: That argument exists to keep related counters in sync.  Some fraction of the time, metrics in the field are by people with hair on fire trying to diagnose a problem. Seeing inconsistent values of related metrics is a distraction that they don't need.  That's also why I added the `isMetricsEnabled` argument to @ServerHandler functions.  (I have removed passing that flag to the `stopConditionalContext()` method, there was enough info with context ==/!= null to use 1 arg only.)

File: infrastructure/src/main/java/org/corfudb/infrastructure/AbstractServer.java (hunk 6 above)
Comment (slfritchie) [inline infrastructure/src/main/java/org/corfudb/infrastructure/AbstractServer.java]: See above (ops "hair on fire") argument for the consistency of related stats. Many times, metrics values will be evaluated by calm people, and the extra arg to make all-or-nothing metrics keeping is wasted effort.  I'd rather add the argument to avoid creating distractions for stressed ops people who don't need more stress.

Comment (dahliamalkhi) [inline infrastructure/src/main/java/org/corfudb/infrastructure/AbstractServer.java]: but the parameter isMetricsEnabled is never used inside any handler.

File: runtime/src/main/java/org/corfudb/runtime/object/CorfuCompileProxy.java (hunk 5 above)
Comment (dahliamalkhi) [inline runtime/src/main/java/org/corfudb/runtime/object/CorfuCompileProxy.java]: I don't understand the explanation.

File: infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java (hunk 12 above)
Comment (slfritchie) [inline infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java]: The overhead of metric objects that are in the registry but never used is quite small.  If that small overhead is a noticeable problem, then we can address it.  If we change things so that metrics are added to the registry only when collection is enabled dynamically, then the annotation-based metrics will need new closures to encapsulate timer references that weren't needed at startup time.  We'd have to run something like the equivalent of `handler = CorfuMsgHandler().generateHandlers()` for each server object when metrics collection status flips from disabled->enabled.

File: infrastructure/src/main/java/org/corfudb/infrastructure/DataStore.java (hunk 8 above)
Comment (slfritchie) [inline infrastructure/src/main/java/org/corfudb/infrastructure/DataStore.java]: It's used once, by ServerContext.  However, you ask about another @Getter annotation added in  SequencerServer that I'll remove (assuming the compiler agrees with my eyes).

File: runtime/src/main/java/org/corfudb/runtime/object/CorfuCompileProxy.java (hunk 5 above)
Comment (slfritchie) [inline runtime/src/main/java/org/corfudb/runtime/object/CorfuCompileProxy.java]: The one case so far of related metrics is the server's operation timer (which is a counter + histogram) and the `counterTokenSum` and `counterToken0` simple counters.  If the timer is updated for any given operation, then the simple counters should always be updated also.  If the timer isn't updated, then counters shouldn't be either.

Comment (slfritchie) [inline runtime/src/main/java/org/corfudb/runtime/object/CorfuCompileProxy.java]: Oops, the above described a consistency example on the server side.  You're pointing to CorfuCompileProxy, sorry.  There's a case there in TXExecuteInner where either counters should either be triggered or not during the loop.

File: infrastructure/src/main/java/org/corfudb/infrastructure/AbstractServer.java (hunk 6 above)
Comment (slfritchie) [inline infrastructure/src/main/java/org/corfudb/infrastructure/AbstractServer.java]: See https://github.com/CorfuDB/CorfuDB/blob/add09105d125dc9730082e66ed7b293066f7da23/infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java#L252-L256

**Comment by slfritchie** [comment]
Ah, I've overlooked the try-with-resources hack.  The `Timer.Context` object implements `Closeable`, hooray, I don't have to invent a new thing.  Sure, I can try it, though there are now many fewer places to use it.

**Comment by slfritchie** [comment]
[...] how to enable/disable metric collection, and where to find the output?

I'll update the wiki topic on stats & metrics.  Previous text was in the old PR, but I ought to add something larger.

**Comment by slfritchie** [comment]
Resolves #234.
